core/include/kernel/panic.h

```diff
@@ -29,9 +29,50 @@
 #define KERNEL_PANIC_H
 
 #include <compiler.h>
+#include <kernel/thread.h>
+#include <trace.h>
 
-#define panic()	__panic(__FILE__, __LINE__, __func__)
+static inline void __panic_prepare(void)
+{
+	(void)thread_mask_exceptions(THREAD_EXCP_ALL);
+}
 
-void __panic(const char *file, int line, const char *func) __noreturn;
+static inline __noreturn void __panic_do(void)
+{
+	while (1)
+		;
+}
+
+#if (CFG_TEE_CORE_DEBUG == 0)
+
+#define panic() \
+	do { \
+		__panic_prepare(); \
+		EMSG_RAW("Panic"); \
+		__panic_do(); \
+	} while (0)
+
+#define panic_trace(...)	panic()
+
+#else
+
+#define panic() \
+	do { \
+		__panic_prepare(); \
+		EMSG_RAW("Panic at %s:%d <%s()>", \
+			 __func__, __LINE__, __func__) ; \
+		__panic_do();	\
+	} while (0)
+
+#define panic_trace(...) \
+	do { \
+		__panic_prepare(); \
```

Comment:
Why aren't we doing this in a single `__panic_xxx()` function? It would save a few bytes and it would be easier to save the old exception mask for inspection in a debugger if desired.